Add bluetooth host keys unit test

ahmedmoheb-nordic opened the pull request:
PR for internal review

rcole-cloud2gnd commented on subsys/bluetooth/host/keys.c:
```diff
@@ -509,3 +509,16 @@ void bt_keys_show_sniffer_info(struct bt_keys *keys, void *data)
 }
 #endif /* defined(CONFIG_BT_LOG_SNIFFER_INFO) */
+
+#ifdef ZTEST_UNITTEST
```
As discussed previously, there are definite concerns about this approach to white-box testing, especially in the context of a security-related module like the key store. Generally, introducing additional interfaces to increase testing coverage of a module is a good thing except for some side effects. The most subtle, but still risky is the potential variation in behavior, especially failure modes, that may occur from having the code that is tested vary from the code that runs. I would be surprised if introducing these methods would ever result in a failure ocurring in the real code but not under test, but it's not impossible. If a failure were to occur it would be from an erroneous or error-prone use of function pointers, offsets, hand-rolled setjmp/longjmp behavior where the presence or absence of code changes behavior. I would consider this a vanishingly small probability, but it is not without precedent. It is also a problem because if it works in unit testing but fails in practice, that leaves a bug, potentially a very difficult to find bug.

The second issue is that of encapsulation violation. This is not just of academic interest. If the bt_keys structure is modified or eliminated in lieu of a different/better approach without other changes to the interface, the unit tests become a dependency that breaks despite no other breakage. More importantly, this invites a future developer to assume access to the key pool is okay and that the interface can be accessed and manipulated simply by enabling the unit-test macro. This could result in introduced dependencies that are hard to eliminate. Again, this is a future-proofing concern and can be worked around, but it leads to the third concern:

The data in the key pool is security-related and any access to the location in memory of the data through an application-level interface opens up more avenues for manipulation and possible future attack vectors. Admittedly, the existing interfaces does not appear hardened against manipulation, but this is another level of access that could be abused. This is less relevant for other interfaces, but an invitation to make 'tricky' fixes is a problem in code designed to be maintained.

Generally if a white-box testing approach is being used and metadata needs to be accessed to show that internal behavior operated correctly, such interfaces changes need to be public and permanent (which may be a hard sell for this and similar modules). In many cases, black-box testing (testing the interface and external behavior without consideration of the internal states) is sufficient, but if the intention is to introduce full coverage testing, that may require specific new behaviors.

I would caution that changes like this (if not already present in the code) have a good chance of being rejected or causing a flame war. We should discuss Nordic's needs and suggest alternatives that are more likely to be approved (or test the waters and find out before introducing a lot of new dependencies).

ahmedmoheb-nordic replied:
We can add that concern on top of items list to be discussed with Nodric team

rcole-cloud2gnd commented on subsys/bluetooth/host/keys.c:
```diff
@@ -509,3 +509,16 @@ void bt_keys_show_sniffer_info(struct bt_keys *keys, void *data)
 #endif /* defined(CONFIG_BT_LOG_SNIFFER_INFO) */
+
+#ifdef ZTEST_UNITTEST
+struct bt_keys *bt_keys_get_key_pool(void)
```
As a matter of interface testing (notwithstanding the previous comments about risks), this should be a pointer to const or const pointer to const to force an explicit cast to enable dangerous manipulation. Ditto subsequent interfaces.

ahmedmoheb-nordic replied:
The point is that some tests need to explicitly clear the list as if it is a fresh start, modifying the pointer to point to constant data will prevent such operation.

rcole-cloud2gnd commented on tests/bluetooth/host/buf/bt_buf_get_evt/src/main.c:
```diff
@@ -0,0 +1,14 @@
+{
+	uint32_t state;
+
+	ztest_run_registered_test_suites(&state);
```
I show this funciton on the deprecated list. How is the Zephyr project treating deprecation in the test suites?

ahmedmoheb-nordic replied:
This one, although it can be found under deprecated ZTest, was used in the `Test Framwork Getting started guide` version 3.1.0 and I just followed that one as it was the latest by that time.

Also, new APIs weren't available previously because the normal CMake build process wasn't available for unit test projects until the one who is now blocking my PR did some modifications.

I preferred not move to the new APIs because it would require some effort while I didn't receive any feedback from Nodic team regarding this point.

I may switch to the new APIs for the next UT module.

rcole-cloud2gnd commented on tests/bluetooth/host/buf/bt_buf_get_evt/src/main.c:
```diff
@@ -0,0 +1,14 @@
+
+void test_main(void)
+{
+	uint32_t state;
```
Considering that the state is theoretically a parameter to the run call and could be used for future uses, it's probably a good idea to initialize the value so that a random value of the stack is not misinterpreted as an initialization value. This is less relevant if removed due to deprecation (see subsequent comment).

rcole-cloud2gnd commented on tests/bluetooth/host/buf/bt_buf_get_evt/src/test_suite_hci_evt_cmd.c:
```diff
@@ -0,0 +1,165 @@
+	TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_STATUS)
+};
+
+BUILD_ASSERT(ARRAY_SIZE(testing_params_lut) == TEST_PARAMETERS_LUT_ROWS_COUNT);
```
I'd need to see if there is pre-existing pattern, but it seems that rather than defiining a row-count as a multiple of the number of testing parameters and then verifying that the fixed value matches an array size that may change, you could define the TEST_PARAMETERS_LUT_ROWS_COUNT as the ARRAY_SIZE of the array which not only will force the assert true (therefore eliminating it), but also ensure that a change to the array does not require a manual and potentially error-prone modification to the count constant.

ahmedmoheb-nordic replied:
Totally agree with you, I initially did it like what you suggested and then I received the following comment

![image](https://user-images.githubusercontent.com/109019690/187082457-334b6468-8393-4433-bfaf-815930b7c9c9.png)

So, I just followed their lead to avoid debate :)

rcole-cloud2gnd replied:
They are correct, you should not be specifying the length of an array with an initializer. This is what I meant:

``

/* LUT containing testing parameters that will be used
 * during each iteration to cover different scenarios
 */
static const struct testing_params testing_params_lut[] = {
	TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_COMPLETE),
	TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_STATUS)
};

/* Rows count equals number of events x 2 */
#define TEST_PARAMETERS_LUT_ROWS_COUNT		ARRAY_SIZE(testing_params_lut) 

```

This ensures that the constant does not become invalid when the testing_params_iut changes size, and no longer requires an assertion for equality because it is equal by definition.

ahmedmoheb-nordic replied:
Ok, I got it now. I think as both approaches have the same effect, we can keep the current style.

Aside from that, why specifying the array size isn't a good practice. 

> you should not be specifying the length of an array with an initializer.

rcole-cloud2gnd replied:
There are cases where specifiying the length and using the initializer is useful, for example if you want to reserve additional zero-initialized space (for example):

`int s_vals[5] = {1, 2};`

You may want to reserve s_vals[2], s_vals[3], and s_vals[4], and this permits that. However, if you define:

```
static const struct testing_params testing_params_lut[CONST_DECLARED_ELSEWHERE] = {
TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_COMPLETE),
TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_STATUS)
};

```
You immediately run into a potential issue if someone else does this:

```
static const struct testing_params testing_params_lut[CONST_DECLARED_ELSEWHERE] = {
TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_COMPLETE),
TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_STATUS),
TEST_PARAM_PAIR_DEFINE(BT_HC_EVT_NEW_THING_I_ADDED)
};

```

That is presumably what the original comment was about. Now you will break the build with the commit unless you remember to (or know to) also change the length declared elswhere (or build locally). Equally, if you remove an entry, there is now an invalid last entry but nothing will warn you:

```
static const struct testing_params testing_params_lut[CONST_DECLARED_ELSEWHERE] = {
TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_COMPLETE)
};

```

The point is that every change to the array that has a side-effect must be made correct by making more than one change. It is always a potential bug or temporary build breakage. In this case you don't have to have multiple interdependent declarations of the array size, because the compiler will select the size of required storage for you.

```
static const struct testing_params testing_params_lut[] = {
TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_COMPLETE),
TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_CMD_STATUS),
TEST_PARAM_PAIR_DEFINE(BT_HC_EVT_NEW_THING_I_ADDED)
};

```

So what you're doing now with the assert is technically correct, but using another independent constant adds an additional step that must be made on any change to the array that will cause a breakage that will need to be analyzed and corrected by someone making a future change to the elements of the array. I would give good odds it will get flagged. But it may not.

ahmedmoheb-nordic replied:
I got your point about why not specifying the array size during the initialization, but I can't see the difference between using the assert like this `BUILD_ASSERT(ARRAY_SIZE(testing_params_lut) == TEST_PARAMETERS_LUT_ROWS_COUNT)`
and the one you are proposing `#define TEST_PARAMETERS_LUT_ROWS_COUNT ARRAY_SIZE(testing_params_lut)`.

Adding/Removing items from the array will require editing the `TEST_PARAMETERS_LUT_ROWS_COUNT` value.

What is the extra step you think will be required using the assert?

rcole-cloud2gnd replied:
"Adding/Removing items from the array will require editing the TEST_PARAMETERS_LUT_ROWS_COUNT value" is the extra step. There is no need to require someone editing this code to understand that subtely.

rcole-cloud2gnd commented on tests/bluetooth/host/buf/bt_buf_get_evt/src/test_suite_hci_evt_default.c:
```diff
@@ -0,0 +1,190 @@
+	TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_AUTH_PAYLOAD_TIMEOUT_EXP)
+};
+
+BUILD_ASSERT(ARRAY_SIZE(testing_params_lut) == TEST_PARAMETERS_LUT_ROWS_COUNT);
```
See prior note on row-count parameter.

rcole-cloud2gnd commented on tests/bluetooth/host/buf/bt_buf_get_evt/src/test_suite_hci_evt_num_completed_packets.c:
```diff
@@ -0,0 +1,161 @@
+	TEST_PARAM_PAIR_DEFINE(BT_HCI_EVT_NUM_COMPLETED_PACKETS),
+};
+
+BUILD_ASSERT(ARRAY_SIZE(testing_params_lut) == TEST_PARAMETERS_LUT_ROWS_COUNT);
```
See note on row-counts.

rcole-cloud2gnd commented on tests/bluetooth/host/keys/testing_common_defs.h:
```diff
@@ -0,0 +1,38 @@
+#define BT_ADDR_LE_5 ((bt_addr_le_t[]){{BT_ADDR_LE_PUBLIC, {{0x0E, 0x89, 0x67, 0x45, 0x23, 0x01}}}})
+
+/** Bluetooth LE device "test" random addresses */
+#define BT_RPA_ADDR_LE_1                                                                           \
```
You may want to refer to BT core specification 7B, 1.3.2.2. the two MSBs of RPA and NRPA random addresses are fixed. If there is code that's handling that properly, you'll probably want to have both valid and invalid address values.

ahmedmoheb-nordic replied:
There is only one check there to check if the address type is random and an RPA

`
	if (!bt_addr_le_is_rpa(addr)) {
		return NULL;
	}
`

The function checks if it is an RPA or not, and the behavior doesn't depend on being NRPA or any other value.
So, I was interested in just turning that check into false and for that I used a public address

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+  message(FATAL_ERROR "This project can only be used with '-DBOARD=unit_testing'.")
+endif()
+
+FILE(GLOB SOURCES src/*.c)
```
Looking at all the cmake projects under keys, they are all very similar and each seems to be implementing only a single test, (add/clear/find) but creating a separate main.c and hence a separate test executable. It's a lot of build overhead and execution overhead and I don't think it's adding much.

It seems that if the keys modules is in use, all of these tests are relevant, in which case there only needs to be one CMakeList.txt, prj.conf, and main.c for all of the tests related to keys. There's a lot of duplication here which could be reduced.

ahmedmoheb-nordic replied:
Your concern is totally valid and I agree with you.
I think we can discuss that point with Nordic team when we meet.

Actually, I think that I combined different tests together at the beginning, then they suggested creating a dedicated UT project for each API.

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,176 @@
+ *  Expected behaviour:
+ *   - All returned references are different from each other
+ */
+void test_no_equal_references(void)
```
I don't understand this test. You appear to be verifying that the existing key pool pointer does not have a duplicate in a list of pointers that are maintained locally and in the same order. That strikes me as a test of the behavior of the test suite rather than the key pool. The current implementation is an array, so it may be reasonable to verify that a returned pointer is NULL or within the range of the array (taking note of encapsulation concerns), but otherwise, testing the integrity of the array can be done with a controlled test to fill it up and check what it does and does not contain. The id has only 256 possible values (currently), so you can do a brute-force test against all ids.

I believe this test is also order-dependent, which is a concern.

ahmedmoheb-nordic replied:
This test mainly verifies that an empty list is filled sequentially in a correct order, so that each reference pointer is moved correctly to the next location which also verifies that no reference pointer was returned twice.

The test iterates over a list with the same size as the keys pool, so boundaries are considered.

On the other hand, NULL checking is performed while adding new items through `bt_keys_get_addr()` (although the name is a bit confusing, this API tries to find a matched record with the same ID and address. If the item exists, it returns the item reference pointer. Otherwise, it tries to add the ID and address pair to the list and then returns the new record reference pointer)

Your concern that this test is order-dependent is totally valid, for example this test doesn't consider the situation when an item in the middle of the list is cleared then reused.

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/main.c:
```diff
@@ -0,0 +1,122 @@
+/* This LUT contains different combinations of ID, Address and key type.
+ * Item in this list will be used to fill keys pool.
+ */
+static const struct id_addr_type testing_id_addr_type_lut[CONFIG_BT_MAX_PAIRED] = {
```
User-configurable parameter (max pairing records in key pool) used with an array with an initializer list. This has a good chance of causing spontaneous compile errors with any implementation-specific selection of max pairing records.

ahmedmoheb-nordic replied:
Regarding defining the array size within the initializer, we agreed before that this isn't a good practice.

But, this parameter `CONFIG_BT_MAX_PAIRED` is configured per UT project to satisfy a specific target and I won't consider it a `User-configurable parameter` as it isn't a generic project.
If someone is going to edit this parameter for a specific UT, he should ensure that the UT is still working.

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/main.c:
```diff
@@ -0,0 +1,122 @@
+
+	{ BT_ADDR_ID_1, BT_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_2, BT_KEYS_IRK },
```
I'll dig into the implications more when I get to the use of this list, but strictly speaking it's an error to have an IRK without a LTK or a CSRK without an LTK, so if these are populating a key-pool record, it may be testing the validity of an invalid case which may be fixed in future.

ahmedmoheb-nordic replied:
Actually, I don't know much about these IRK, LTK and CSRK and the relations between them.
The implementation of `bt_keys_add_type` is pretty simple, it just sets what you pass to it.

I'd say that in a Test-Driven-Development, the test case should be designed to consider such relations.
I'm not sure if I should be restricted to the current implementation while designing test case, or I should go beyond the implementation.
Also, you were able to catch such thing because you have the right knowledge, but I didn't know about it.

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/main.c:
```diff
@@ -0,0 +1,122 @@
+/*
```
Out of curiosity, why is the most recent parent commit for these tests from 2016? What version of the code is being targeted? A Bluetooth host from 2016 is not necessarily qualifiable for new designs (it may be, but it's a risk).

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/main.c:
```diff
@@ -0,0 +1,122 @@
+	keys_fill_pool();
+
+	ztest_test_suite(bt_keys_add_type_verify_type_value,
+			 LISTIFY(CONFIG_BT_MAX_PAIRED, REGISTER_SETUP_TEARDOWN, (,),
```
I don't understand the value of this parameterization over set of max pairings. Firstly, the set of possible combinations to test could easily be larger or smaller than CONFIG_BT_MAX_PAIRED, secondly any attempt to re-order the tests will fail because this is simply repeating the same configuration CONFIG_BT_MAX_PAIRED times against params_it, which is not under ztest's control. It should be equivalent and simpler to have one test that simply iterates the scenarios defined by testing_id_addr_type_lut. I suppose you would have any assert terminate before all tests run, which could require multiple iterations to identify whether the tests passes (after a change which breaks multiple behaviors), but that seems a design limitation of ztest (no EXPECT_xxx as in gtest or other soft-assert mechanism).

ahmedmoheb-nordic replied:
This is an old debate about what is supported by ZTest?

The problem was "How can I run the same test case with a set of input parameters?"

There are some limitations by ZTest

1. Test case function can't have any arguments and it must have the form `void test_name(void)`
2. You can run the UT project with different parameters by adding multiple records to the `testcase.yaml`.
    But, it doesn't support to define a list of parameters within a certain record.
    You will end up with a lot of records equals to the number of parameters you need to run

- One way, as you mentioned, was to run a single test case and iterate over the items using a loop.
  The problem was that the output report will include a single record for that test case with the result pass or fail for all parameters, so you won't be able to recognize which iterations failed.

- The second approach, which is currently used, was to duplicate the test case and depend on a global iterator to select the parameters to be tested.

Althoug I was in favor of the second approach by that time and they agreed on it, I'm not any more.
It's not easy to follow and harder to re-order.

Actually, now I'm looking at the use of new ZTest APIs and I'm trying to look for another way to do it away from duplicating test cases.

ahmedmoheb-nordic replied:
> I don't understand the value of this parameterization over set of max pairings.

`CONFIG_BT_MAX_PAIRED` is just used to set number of iterations

rcole-cloud2gnd replied:
Except that CONFIG_BT_MAX_PAIRED is 16 (I think), but could probably be set as little as 2. The combinations aren't important to the current implementation of bt_keys_add_type, but if they were, you could have lots more combinations. Given a static and random address, LTK or 256-bit LTK, presence or absence of local IRK, presence or absence of local CSRK, presence or absence of remote IRK, presence or absence of remote CSRK... that's 64 combinations already.

ahmedmoheb-nordic replied:
I already changed that based on your previous comment.
Now I'm iterating of the list in one single test case and added a new one to test the value of `BT_KEYS_ALL`

I will consider applying the concept of iterations with the upcoming modules.
I'm not in favor of duplicating test cases and depend on global parameters anymore

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/main.c:
```diff
@@ -0,0 +1,122 @@
+	int type = current_params_vector->type;
+
+	/* 'params_it' is incremented before running the test case */
+	key_ref = returned_keys_refs[params_it - 1];
```
Isn't this returned_keys_refs equivalent to the result of bt_keys_get_addr() from the keys_fill_pool call? Isn't the value of elements in the returned_keys_ref array going to be indempotent between the keys_fill_pool() and the test iteration. If so, why not call bt_keys_get_addr here to add the entries as you use them, assert on failure and then directly use the result? That ensures that if tests are run out of order that entries are added in different orders and you better emulate any side-effects.

ahmedmoheb-nordic replied:
Excellent remark, it made my life easier :)

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/main.c:
```diff
@@ -0,0 +1,122 @@
+	zassert_true(check_key_pool_is_empty(),
+		     "List isn't empty, make sure to run this test just after a fresh start");
+
+	for (size_t it = 0; it < ARRAY_SIZE(testing_id_addr_type_lut); it++) {
```
I would suggest since this module is about the behavior of bt_keys_add_type, this specific list of pool values isn't really relevant. bt_keys_add_type as its written updates the value of a mask in a bt_keys structure, irrespective of where that structure resides. It's reasonable to ensure that a bt_keys value in the pool changes when bt_keys_add_type is used to future-proof the tests, but you could define a bt_keys on the stack and verify that the mask changes correctly without any of this setup, teardown or iteration. The coverage impact of the test (today) would be the same, assuming you are not relying on this additional testing to fill in blanks on bt_keys_get_addr testing. If so, I'd move those behaviors to the bt_keys_get_addr test.

ahmedmoheb-nordic replied:
Good one

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/main.c:
```diff
@@ -0,0 +1,122 @@
+void test_set_type_value_correctly(void)
+{
+	struct bt_keys *key_ref;
+	int type = current_params_vector->type;
```
Rather than a separate test run for each viable type, I might recommend that the valid mask values in BT_KEYS_ALL are all tested in one pass or that all valid bits are checked in case some flags are made invalid in future.

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_add_type/src/test_suite_add_type_invalid_inputs.c:
```diff
@@ -0,0 +1,37 @@
+void test_null_key_reference(void)
+{
+	expect_assert();
+	bt_keys_add_type(NULL, 0x00);
```
If this bt_keys_add_type is a fake, this is not testing the NULL case. If it's not a fake, I would expect this to actually result in a segmentation fault or protection violation. Has this been executed?

I might recommend that you consider adding an assertion to bt_keys_add_type() for an invalid pointer, though I don't know if that's in scope for this project.

rcole-cloud2gnd replied:
Also, is a valid key_ref and zero mask tested? It appears to have no effect, which may become a relied-upon behavior.

ahmedmoheb-nordic replied:
> I might recommend that you consider adding an assertion to bt_keys_add_type() for an invalid pointer, though I don't know if that's in scope for this project.

Actually, this is a test for the assert to be triggered (which I added).

> Also, is a valid key_ref and zero mask tested? It appears to have no effect, which may become a relied-upon behavior.

I didn't think about valid key with zero mask, but it seems to be a valid case

rcole-cloud2gnd replied:
I'm not clear on why an assert would be trigged. What triggers it? Keep in mind, if the original code is running, the current implementation is:

`keys->keys |= type;`

If the keys pointer is null, this is accessing memory at an offset from 0x0:

`*(uint16_t *)(((uint8_t *)NULL)+offsetof(struct bt_keys, keys)) = new_val;`

That's not going to assert, that's going to hard-fault (on most platforms).

How is fatal-errors handling configured?

ahmedmoheb-nordic replied:
Sorry, may be I wasn't clear enough.

I already modified the `keys.c` a little bit to include such asserts and you can find the modifications here [Enhance bluetooth host keys](https://github.com/zephyrproject-rtos/zephyr/pull/49596).

The UTs projects are dependent on these modifications and that's why CI checks are always failing because the modifications need to be merged first.

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_clear/src/main.c:
```diff
@@ -0,0 +1,210 @@
+	const bt_addr_le_t *addr_to_find = current_params_vector->addr;
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		ztest_test_skip();
```
CONFIG_BT_SETTINGS controls whether the flash settings is updated, but it doesn't stop the keys record from being cleared or the bt_id_del() call being called if the key state includes BT_KEYS_ID_ADDED.

ahmedmoheb-nordic replied:
Yes, that's correct.

The `testcase.yaml` for this project contains 2 sets of configurations which will result in executing the project once for each set.

The point is that there is another version of that test to be executed when `CONFIG_BT_SETTINGS` is enabled.
The other one make some more checks and can be found here `bt_keys_clear/src/test_suite_bt_settings.c`

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_clear/src/main.c:
```diff
@@ -0,0 +1,210 @@
+
+	bt_keys_clear(key_ref_to_clear);
+
+	expect_single_call_bt_id_del(key_ref_to_clear);
```
I would be concerned that this approach (having one function that tests that a flag results in a behavior and a completely different function that checks that the absence of that flag has a different behavior could result in a combinatoric explosion of cases as you get into more complex modules. If you configure your cases with a mix of parameters and use whether the flag was set before clear to determine whether the mock of bt_id_del was called you could simplify this quite a bit.

ahmedmoheb-nordic replied:
Yes, your point is totally valid, till now I'm trying to avoid adding a lot of checks to make the test as concise as possible and only testing one path.

Also, they previously complained about adding branching inside the test case, but it's hard to avoid them when the behavior is more dependent on different configuration parameters

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_clear/src/main.c:
```diff
@@ -0,0 +1,210 @@
+	expect_not_called_bt_id_del();
+
+	/* Verify that memory was cleared */
+	zassert_mem_equal(key_ref_to_clear, &empty_key, sizeof(struct bt_keys),
```
Since bt_keys is a user-facing structure, there's no reason I can't add additional storage after it and cast a bt_keys type (for use in some functions). Does bt_keys_clear leave that memory alone?

```
struct bt_keys *my_keys_data = malloc(sizeof(struct bt_keys) + sizeof(struct bt_my_data));
init_my_key(my_keys_data);
bt_keys_clear(my_keys_data);

```

Does it consider that memory invalid if it's not in the key pool?

ahmedmoheb-nordic replied:
> Does bt_keys_clear leave that memory alone?

Yes!

> Does it consider that memory invalid if it's not in the key pool?

No, I already thought about this one whether the function should check if the input key reference is a valid one with respect to the keys pool boundaries or not.

Such test cases will require some modifications to the original implementation, Do you think we should do that?

rcole-cloud2gnd commented on tests/bluetooth/host/keys/bt_keys_clear/src/main.c:
```diff
@@ -0,0 +1,210 @@
+			  "Key content wasn't cleared by 'bt_keys_clear()'");
+
+	/* Ensure that item doesn't exist in the keys pool after calling bt_keys_clear() */
+	find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
```
If the key was in the key pool but didn't have BT_KEYS_ID_ADDED set in the state, it's presumably not going to be deleted in this case, so shouldn't this always fail is the previous bt_keys_find_addr suceeded?

ahmedmoheb-nordic replied:
Yes, the second check is checking if the return value is `NULL`

```
/* Ensure that item doesn't exist in the keys pool after calling bt_keys_clear() */
find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
zassert_true(find_returned_ref == NULL,
		     "bt_keys_find_addr() returned a non-NULL reference");
```

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add mocks for keys.c

Add required mocks to be able to compile and test /bluetooth/host/keys.c

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add UT for bt_keys_get_addr()

Unit test project for bt_keys_get_addr().
This part of subsys/bluetooth/host/buf.c unit testing

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>
